### PR TITLE
Private dns support

### LINKED ISSUE #47: Issue when managing private DNS zone
Hey,

We are trying to manage our private DNS zone with octo.

We already have created our main private zone on Azure which looks like env.cloudprovider.company.internal.

Our problem is when we create a record using this module. It create a complete new DNS Zone (and not a private one) which looks like record.env.cloudprovider.company.internal instead of using the existing private zone

Are we doing something wrong or the module just don't support private DNS zone ?

Thanks

### PR DESCRIPTION
**Pull request by ross**
This adds a commit onto https://github.com/octodns/octodns-azure/pull/49 which refactors things to have private DNS be a separate class, `AzurePrivateProvider` rather than a flag to the existing class and a bunch of if's. This generally is better OO practice, but it's been long enough since I've looked at that stuff in depth that I don't remember the name/rule for doing this :grin:

Once you all have 👀 this and 👞 tires we can move forward on with it.

/cc https://github.com/octodns/octodns-azure/pull/49 which this adds onto @alexnl022 @jpuchois 
/cc Fixes #47 
/cc @viranch

### CODE REVIEW

File: octodns_azure/__init__.py [hunk 1]
```diff
@@ -1754,3 +1676,258 @@ def _apply(self, plan):
...
+
+        Example config file with variables:
+            "
+            ---
+            providers:
+              config:
+                class: octodns.provider.yaml.YamlProvider
+                directory: ./config (example path to directory of zone files)
+              azuredns:
+                class: octodns.provider.azuredns.AzureProvider
```
Comment (viranch) [inline octodns_azure/__init__.py:1711]:
```suggestion
                class: octodns_azure.AzureProvider
```

File: octodns_azure/__init__.py [hunk 2]
```diff
@@ -1754,3 +1676,258 @@ def _apply(self, plan):
...
```
Comment (viranch) [inline octodns_azure/__init__.py:1836]:
```suggestion
                class: octodns_azure.AzurePrivateProvider
```

File: octodns_azure/__init__.py [hunk 3]
```diff
@@ -1754,3 +1676,258 @@ def _apply(self, plan):
...
+        delete = self.dns_client.record_sets.delete
+        delete(resource_group, zone_name, relative_record_set_name, record_type)
+
+
+class AzurePrivateProvider(AzureBaseProvider):
+    '''
+    Azure DNS Provider
+
+    azuredns.py:
+        class: octodns.provider.azuredns.AzureProvider
```
Comment (viranch) [inline octodns_azure/__init__.py:1811]:
```suggestion
        class: octodns_azure.AzurePrivateProvider
```

File: octodns_azure/__init__.py [hunk 4]
```diff
@@ -1754,3 +1676,258 @@ def _apply(self, plan):
...
                 continue
             getattr(self, f'_apply_{class_name}')(change)
+
+
+class AzureProvider(AzureBaseProvider):
+    '''
+    Azure DNS Provider
+
+    azuredns.py:
+        class: octodns.provider.azuredns.AzureProvider
```
Comment (viranch) [inline octodns_azure/__init__.py:1686]:
```suggestion
        class: octodns_azure.AzureProvider
```

**Review by viranch — COMMENTED** [review]
Had some time to test this out IRL. The only roadblock was lack of NS records in the private zone, which causes the following exception:

```
  File "/home/viranch/code/octodns-tests/pvt-dns/venv/lib/python3.10/site-packages/octodns_azure/__init__.py", line 629, in _check_zone
    self._required_root_ns_values[name] = set(zone.name_servers)
AttributeError: 'NoneType' object has no attribute 'name_servers'
```

Looks like `PrivateZone` isn't supposed to have the `name_servers` attribute: https://learn.microsoft.com/en-us/python/api/azure-mgmt-privatedns/azure.mgmt.privatedns.models.privatezone?view=azure-python

I was able to work around this by setting `SUPPORTS_ROOT_NS = False` under `AzurePrivateProvider`. After that, the zone was created with all the records as expected.

It also likely makes sense to disable `SUPPORTS_DYNAMIC`. While it works in my testing, the resulting Traffic Managers are no different, in that they're resolvable by the world which is opposite of private DNS. So unless there's a use-case, I'd start by not supporting them.

Rest looks good with some minor suggestions below.

**Comment by ross** [comment]
I believe the nameserver and dynamic/trafficmanager stuff has been addressed. The traffic manager bit turned out to be a little more involved than just adding `SUPPORTS_DYNAMIC = false` once I got to looking at it as it seemed much cleaner to pull all of that stuff out of the base class and have it only exist in the main provider. Also removed a bunch of dyanmic tests that were hitting the private dns provider. This should be good for another quick set of :eyes: before shipping.

**Comment by viranch** [comment]
It's probably better to actually move the dynamic and root NS code out of base into the AzureProvider. It's likely more involved, so I can do it in a separate PR.

**Comment by ross** [comment]
> It's probably better to actually move the dynamic and root NS code out of base into the AzureProvider. It's likely more involved, so I can do it in a separate PR.

I moved the dynamic/tm stuff out in this PR already. I didn't move the root NS stuff out, but I suspect it would be less involved.

**Comment by viranch** [comment]
Moved the rest of the dynamic stuff out of base as well as the root NS stuff. Ran basic tests against my Azure account and seems to be functional. This is ready to ship it you're happy with my commits @ross

**Review by viranch — APPROVED** [review]

**Review by ross — COMMENTED** [review]
:cool: there was a lot more that was related to the dynamic stuff than realized not being familiar with it.

**Pull request merged**